### PR TITLE
The ESM problem - executing the controller itself as an ESM module

### PR DESCRIPTION
**Pull request by foxriver76**

### CODE REVIEW

File: packages/common/src/lib/common/logger.ts [hunk 1]
```diff
@@ -1,28 +1,34 @@
...
+
+// eslint-disable-next-line unicorn/prefer-module
+const require = createRequire(import.meta.url || 'file://' + __dirname);
 
 const hostname = tools.getHostName();
 
 let SysLog: typeof Syslog | undefined;
 try {
-    // eslint-disable-next-line @typescript-eslint/no-var-requires
     SysLog = require('winston-syslog').Syslog;
```
Comment (foxriver76) [inline packages/common/src/lib/common/logger.ts:25]: We cannot use top-level await if we transpile to cjs with https://github.com/AlCalzone/esm2cjs, hence I kept it require calls

File: packages/common/src/lib/common/tools.ts [hunk 2]
```diff
@@ -336,8 +353,8 @@ function getAppName(): string {
     return 'iobroker';
 }
 
+export const appNameLowerCase = 'iobroker';
```
Comment (foxriver76) [inline packages/common/src/lib/common/tools.ts:356]: `getAppName` uses `isDevInstallation` which uses `lowercaseAppName` leading to a loop, hence I hardcoded it.

File: packages/common/src/lib/common/tools.ts [hunk 3]
```diff
@@ -1546,9 +1560,7 @@ export function getAdapterDir(adapter: string): string | null {
             adapterPath = path.join(getControllerDir(), 'node_modules', possibility);
         } else {
             try {
-                adapterPath = require.resolve(possibility, {
-                    paths: getDefaultRequireResolvePaths(module)
```
Comment (foxriver76) [inline packages/common/src/lib/common/tools.ts:1550]: I have not found a way to port the `paths` to esm, I hope the require resolve works anyway in the occuring scenarios.

File: packages/common/src/lib/common/tools.ts [hunk 4]
```diff
@@ -3143,6 +3153,22 @@ export function pipeLinewise(input: NodeJS.ReadableStream, output: NodeJS.Writab
...
+ *
+ * @param adapter name of the adapter like hm-rpc
+ */
+export async function isAdapterEsmModule(adapter: string): Promise<boolean> {
+    const adapterDir = getAdapterDir(adapter);
+    if (!adapterDir) {
+        throw new Error(`Could not find adapter dir of ${adapter}`);
+    }
+
+    const packJson = await fs.readJSON(path.join(adapterDir, 'package.json'), { encoding: 'utf-8' });
```
Comment (foxriver76) [inline packages/common/src/lib/common/tools.ts:3167]: in theory the adapter main file could lie in another dir having a separate package.json with a separate type argument, but I think nobody should do this..

File: packages/controller/controller.js [hunk 5]
```diff
@@ -6,4 +6,4 @@
  *
  */
 
-require('./build/cjs/main').init();
+import('./build/esm/main.js').then(res => res.init());
```
Comment (foxriver76) [inline packages/controller/controller.js:9]: no top level await so these calls are normal promises

**Review by Apollon77 — APPROVED** [review]
beautiful :-)

**Pull request merged**